tools/autograd/gen_variable_type.py

```diff
@@ -84,18 +286,79 @@ def create_autograd_functions(top_env, declarations):
     These contain the auto-generated subclasses of torch::autograd::Function
     for each every differentiable torch function.
     """
-    # function_definitions = top_env['autograd_function_definitions']
-    # function_declarations = top_env['autograd_function_declarations']
+    function_definitions = top_env['autograd_function_definitions']
+    function_declarations = top_env['autograd_function_declarations']
+    py_function_initializers = top_env['py_autograd_function_initializers']
+
+    def group_by_name(options):
+        res = {}
+        for option in options:
+            name = option['name']
+            if name not in res:
+                res[name] = []
+            res[name].append(option)
+        return res
+
+    for name, options in group_by_name(declarations).items():
+        for i, option in enumerate(options):
+            name = option['name']
+            option['op'] = name[0].upper() + name[1:] + 'Backward'
+            if len(options) > 1:
+                option['op'] += str(i)
+
+    def process_function(op):
+        if op['fallthrough']:
+            return
+
+        saved_variables = []
+        for arg in op['saved']:
+            name = arg['name']
+            if arg['type'] == 'Tensor':
+                saved_variables.append('SavedVariable {}_;'.format(name))
+            elif arg['type'] == 'IntList':
+                saved_variables.append('std::vector<int64_t> {};'.format(name))
+            else:
+                saved_variables.append('{} {};'.format(arg['type'], name))
+        op['saved_variables'] = saved_variables
 
-    def process_function(option):
-        # TODO: generate autograd::Function classes for each function specified
-        # in derivatives.yaml
-        pass
+        body = []
+        body.append('auto& grad = inputs[0];')
+
+        def unpack_args(derivative):
+            unpack = []
+            for arg in op['saved']:
+                if arg['type'] == 'Tensor':
+                    name = arg['name']
+                    unpack.append('auto {} = {}_.unpack();'.format(name, name))
+            return unpack
+
+        i = 0
+        for arg in op['python_arguments']:
+            derivative = arg.get('derivative')
+            if derivative is None:
+                continue
+            body.append(DERIVATIVE.substitute({
+                'unpack_save_variables': unpack_args(derivative),
+                'i': i,
+                'derivative': derivative,
+            }))
+            i += 1
```

Comment:
We don't support things that return non-differentiable outputs right now, do we?